File: world/tile.py
"""地图瓦片。"""
import pygame

import config
from entities.base import Entity
from utils.helpers import load_image


class Tile(Entity):
    """地图瓦片基类。"""

    def __init__(self, x, y, image_name, color, solid=True, destructible=False, blocks_bullets=True):
        image = load_image(image_name, (config.TILE_SIZE, config.TILE_SIZE), color)
        super().__init__(x, y, image)
        self.solid = solid
        self.destructible = destructible
        self.blocks_bullets = blocks_bullets
        self.gx = x // config.TILE_SIZE
        self.gy = y // config.TILE_SIZE

    def take_damage(self):
        """被子弹命中。"""
        if self.destructible:
            self.kill()


class Brick(Tile):
    def __init__(self, x, y):
        super().__init__(x, y, "brick.png", config.COLOR_BROWN, solid=True, destructible=True, blocks_bullets=True)


class Steel(Tile):
    def __init__(self, x, y):
        super().__init__(x, y, "steel.png", config.COLOR_STEEL, solid=True, destructible=False, blocks_bullets=True)


class Water(Tile):
    def __init__(self, x, y):
        super().__init__(x, y, "water.png", config.COLOR_WATER, solid=True, destructible=False, blocks_bullets=False)


class Grass(Tile):
    def __init__(self, x, y):
        super().__init__(x, y, "grass.png", config.COLOR_GRASS, solid=False, destructible=False, blocks_bullets=False)


class Eagle(Tile):
    def __init__(self, x, y):
        super().__init__(x, y, "eagle.png", config.COLOR_YELLOW, solid=True, destructible=True, blocks_bullets=True)

    def take_damage(self):
        """基地被毁，触发游戏结束。"""
        self.kill()
        # 游戏结束标志由 PlayState 通过检测 Eagle 是否存活来处理

File: js/world/tile.js
"use strict";

var CONFIG = require('../config');
var Entity = require('../entities/base');

function Tile(x, y, drawFn, color, options) {
  options = options || {};
  Entity.call(this, x, y, CONFIG.TILE_SIZE, CONFIG.TILE_SIZE);
  this.drawFn = drawFn;
  this.color = color;
  this.solid = options.solid !== undefined ? options.solid : true;
  this.destructible = options.destructible || false;
  this.blocksBullets = options.blocksBullets !== undefined ? options.blocksBullets : true;
  this.gx = Math.floor(x / CONFIG.TILE_SIZE);
  this.gy = Math.floor(y / CONFIG.TILE_SIZE);
}

Tile.prototype = Object.create(Entity.prototype);
Tile.prototype.constructor = Tile;

Tile.prototype.takeDamage = function () {
  if (this.destructible) {
    this.kill();
  }
};

Tile.prototype.draw = function (ctx) {
  if (!this.alive) {
    return;
  }
  if (this.drawFn) {
    this.drawFn(ctx, this.x, this.y, this.w);
  } else {
    ctx.fillStyle = this.color;
    ctx.fillRect(this.x, this.y, this.w, this.h);
  }
};

function Brick(x, y, assets) {
  var drawFn = function (ctx, dx, dy, size) {
    assets.drawBrick(ctx, dx, dy, size);
  };
  Tile.call(this, x, y, drawFn, CONFIG.COLORS.BROWN, { destructible: true });
}
Brick.prototype = Object.create(Tile.prototype);
Brick.prototype.constructor = Brick;

function Steel(x, y, assets) {
  var drawFn = function (ctx, dx, dy, size) {
    assets.drawSteel(ctx, dx, dy, size);
  };
  Tile.call(this, x, y, drawFn, CONFIG.COLORS.STEEL);
}
Steel.prototype = Object.create(Tile.prototype);
Steel.prototype.constructor = Steel;

function Water(x, y, assets) {
  var drawFn = function (ctx, dx, dy, size) {
    assets.drawWater(ctx, dx, dy, size);
  };
  Tile.call(this, x, y, drawFn, CONFIG.COLORS.WATER, { blocksBullets: false });
}
Water.prototype = Object.create(Tile.prototype);
Water.prototype.constructor = Water;

function Grass(x, y, assets) {
  var drawFn = function (ctx, dx, dy, size) {
    assets.drawGrass(ctx, dx, dy, size);
  };
  Tile.call(this, x, y, drawFn, CONFIG.COLORS.GRASS, {
    solid: false,
    blocksBullets: false
  });
}
Grass.prototype = Object.create(Tile.prototype);
Grass.prototype.constructor = Grass;

function Eagle(x, y, assets) {
  var drawFn = function (ctx, dx, dy, size) {
    assets.drawEagle(ctx, dx, dy, size);
  };
  Tile.call(this, x, y, drawFn, CONFIG.COLORS.YELLOW, { destructible: true });
}
Eagle.prototype = Object.create(Tile.prototype);
Eagle.prototype.constructor = Eagle;

module.exports = {
  Tile: Tile,
  Brick: Brick,
  Steel: Steel,
  Water: Water,
  Grass: Grass,
  Eagle: Eagle
};
